paimon-core/src/main/java/org/apache/paimon/operation/FileStoreCommitImpl.java

```diff
@@ -876,41 +923,54 @@ boolean tryCommitOnce(
+            mergeAfterManifests =
                     ManifestFileMerger.merge(
-                            oldMetas,
+                            mergeBeforeManifests,
                             manifestFile,
                             manifestTargetSize.getBytes(),
                             manifestMergeMinCount,
                             manifestFullCompactionSize.getBytes(),
                             partitionType,
-                            manifestReadParallelism));
-            previousChangesListName = manifestList.write(newMetas);
+                            manifestReadParallelism);
+            baseManifestList = manifestList.write(mergeAfterManifests);
 
             // the added records subtract the deleted records from
-            long deltaRecordCount = recordCountAdd(tableFiles) - recordCountDelete(tableFiles);
+            long deltaRecordCount = recordCountAdd(deltaFiles) - recordCountDelete(deltaFiles);
             long totalRecordCount = previousTotalRecordCount + deltaRecordCount;
 
-            // write new changes into manifest files
-            List<ManifestFileMeta> newChangesManifests = manifestFile.write(tableFiles);
-            newMetas.addAll(newChangesManifests);
-            newChangesListName = manifestList.write(newChangesManifests);
+            // write new delta files into manifest files
+            deltaManifestList =
+                    retryResult == null
+                            ? manifestList.write(manifestFile.write(deltaFiles))
+                            : retryResult.deltaManifestList;
 
             // write changelog into manifest files
             if (!changelogFiles.isEmpty()) {
-                changelogMetas.addAll(manifestFile.write(changelogFiles));
-                changelogListName = manifestList.write(changelogMetas);
+                changelogManifestList =
+                        retryResult == null
+                                ? manifestList.write(manifestFile.write(changelogFiles))
+                                : retryResult.changelogManifestList;
             }
 
             // write new index manifest
-            String indexManifest =
-                    indexManifestFile.writeIndexFiles(
-                            previousIndexManifest, indexFiles, bucketMode);
-            if (!Objects.equals(indexManifest, previousIndexManifest)) {
-                newIndexManifest = indexManifest;
+            boolean rewriteIndexManifest = true;
+            if (retryResult != null) {
+                if (Objects.equals(oldIndexManifest, retryResult.oldIndexManifest)) {
+                    rewriteIndexManifest = false;
+                    indexManifest = retryResult.newIndexManifest;
+                    LOG.info("Reusing index manifest {} for retry.", indexManifest);
+                } else {
+                    cleanIndexManifest(retryResult.oldIndexManifest, retryResult.newIndexManifest);
+                }
+            }
```

Comment:
NIT: Use a big `if... else...` block, instead of two `?:` and an `if... else...`